crates/miden-lib/asm/miden/active_note.masm

```diff
@@ -0,0 +1,367 @@
+        call.wallet::receive_asset
+        # => [pad(16), ptr, end_ptr]
+
+        # clean the stack after call
+        dropw dropw dropw
+        # => [EMPTY_WORD, ptr, end_ptr]
+
+        # increment the pointer and continue looping if ptr != end_ptr
+        movup.4 add.4 dup dup.6 neq
+        # => [should_loop, ptr+4, EMPTY_WORD, end_ptr]
+    end
+    # => [ptr', EMPTY_WORD, end_ptr]
+
+    # clear the stack
+    drop dropw drop
+    # => []
+end
+
+# HELPER PROCEDURES
+# =================================================================================================
+
+#! Writes the note inputs stored in the advice map to the memory specified by the provided 
+#! destination pointer.
+#!
+#! Inputs:
+#!   Operand stack: [NOTE_INPUTS_COMMITMENT, num_inputs, dest_ptr]
+#!   Advice map: {
+#!     NOTE_INPUTS_COMMITMENT: [[INPUT_VALUES]]
+#!   }
+#! Outputs:
+#!   Operand stack: [num_inputs, dest_ptr]
+proc.write_inputs_to_memory
+    # load the inputs from the advice map to the advice stack
+    adv.push_mapvaln
+    # OS => [NOTE_INPUTS_COMMITMENT, num_inputs, dest_ptr]
+    # AS => [advice_num_inputs, [INPUT_VALUES]]
+
+    # move the number of inputs obtained from advice map to the operand stack
+    adv_push.1 dup.5
+    # OS => [num_inputs, advice_num_inputs, NOTE_INPUTS_COMMITMENT, num_inputs, dest_ptr]
+    # AS => [[INPUT_VALUES]]
+
+    # Validate the note inputs length. Round up the number of inputs to the next multiple of 8: that
+    # value should be equal to the length obtained from the `adv.push_mapvaln` procedure.
+    u32divmod.8 neq.0 add mul.8
+    # OS => [rounded_up_num_inputs, advice_num_inputs, NOTE_INPUTS_COMMITMENT, num_inputs, dest_ptr]
+    # AS => [[INPUT_VALUES]]
+
+    assert_eq.err=ERR_NOTE_INVALID_NUMBER_OF_INPUTS
+    # OS => [NOTE_INPUTS_COMMITMENT, num_inputs, dest_ptr]
+    # AS => [[INPUT_VALUES]]
+
+    # calculate the number of words required to store the inputs
+    dup.4 u32divmod.4 neq.0 add
+    # OS => [num_words, NOTE_INPUTS_COMMITMENT, num_inputs, dest_ptr]
+    # AS => [[INPUT_VALUES]]
+
+    # round up the number of words to the next multiple of 2
+    dup is_odd add
+    # OS => [even_num_words, NOTE_INPUTS_COMMITMENT, num_inputs, dest_ptr]
+    # AS => [[INPUT_VALUES]]
+
+    # prepare the stack for the `pipe_preimage_to_memory` procedure
+    dup.6 swap
+    # OS => [even_num_words, dest_ptr, NOTE_INPUTS_COMMITMENT, num_inputs, dest_ptr]
+    # AS => [[INPUT_VALUES]]
+
+    # write the inputs from the advice stack into memory
+    exec.mem::pipe_preimage_to_memory drop
+    # OS => [num_inputs, dest_ptr]
+    # AS => []
+end
+
+#! Extracts the sender ID from the provided metadata word.
+#!
+#! Inputs:  [METADATA]
+#! Outputs: [sender_id_prefix, sender_id_suffix]
+#!
+#! Where:
+#! - METADATA is the metadata of some note.
+#! - sender_{prefix,suffix} are the prefix and suffix felts of the sender ID of the note which 
+#!   metadata was provided.
+proc.extract_sender_from_metadata
```

Comment:
Not for this PR, but we may want to move this to the `miden::note` module as we will want to use it for input notes as well. Let's create an issue to add `miden::input_note::get_sender` procedure and then as a part of addressing it, we'll also relocate this procedure.